lib/spans/base-span-streamer.js

```diff
@@ -0,0 +1,135 @@
+      logger.info('Span streamer connected')
+      this.stream = stream
+      this._writable = true
+      this.sendQueue()
+    })
+
+    this.connection.on('disconnected', () => {
+      logger.info('Span streamer disconnected')
+      this.stream = null
+      this._writable = false
+    })
+  }
+
+  addToQueue() {
+    throw new Error('addToQueue is not implemented')
+  }
+
+  /* Accepts a span and either writes it to the stream, queues it to be sent,
+   * or drops it depending on stream/queue state
+   */
+  write(span) {
+    this._metrics.getOrCreateMetric(NAMES.SEEN).incrementCallCount()
+
+    // If not writeable (because of backpressure) queue the span
+    if (!this._writable) {
+      if (this.spans.length < this.queue_size) {
+        this.addToQueue(span)
+        return
+      }
+
+      // While this can be directionally calculated between seen/sent the
+      // queue makes that a bit more disconnected. This will be a bit more specific.
+      this._metrics.getOrCreateMetric(NAMES.DROPPED).incrementCallCount()
+
+      // If the queue is full drop the span
+      logger.infoOncePer(
+        'SPAN_DROP_MSG', // key for the OncePer
+        SPAN_DROP_MSG_INTERVAL_MS,
+        SPAN_DROP_MSG
+      )
+    }
+  }
+
+  /**
+   *  Sends the data(spans or span) over the stream. Spans are only sent here if the stream is
+   *  in a writable state. If the stream becomes unwritable after sending the
+   *  span, a drain event handler is setup to continue writing when possible.
+   *
+   * @param {*} data spans or span
+   * @param {number} [spanLen=1] number of spans sent in a stream(defaults to 1)
+   */
+  send(data, spanLen = 1) {
+    // false indicates the stream has reached the highWaterMark
+    // and future writes should be avoided until drained. written items,
+    // including the one that returned false, will still be buffered.
+    this._writable = this.stream.write(data)
+    this._metrics.getOrCreateMetric(NAMES.SENT).incrementCallCount(spanLen)
+
+    if (!this._writable) {
+      const waitDrainStart = Date.now()
+      const onDrain = this.drain.bind(this, waitDrainStart)
+      this.stream.once('drain', onDrain)
+    }
+  }
+
+  /**
+   *  Drains the span queue that built up when the connection was
+   *  back-pressured or disconnected. `waitDrainStart` is when the stream
+   *  initially blocked, used to time how long the stream was blocked. If this
+   *  is not defined, it is assumed this is being called after a reconnect,
+   *  and the metric is not used.
+   *
+   * @param {number} waitDrainStart time that drain started
+   */
+  drain(waitDrainStart) {
+    // Metric can be used to see how frequently completing drains as well as
+    // average time to drain from when we first notice.
+    const drainCompleted = Date.now()
+    const drainDurationMs = drainCompleted - waitDrainStart
+    this._metrics.getOrCreateMetric(NAMES.DRAIN_DURATION).recordValue(drainDurationMs / 1000)
+
```

Comment:
We usually measure durations metrics in milliseconds. Why the division by 1000 here?